ENH Remove unnecessary OOB computation when n_more_estimators == 0

choo8 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Fixes #20435, based on the comments by @NicolasHug (https://github.com/scikit-learn/scikit-learn/issues/20435#issuecomment-872835169)

#### What does this implement/fix? Explain your changes.
Removes unnecessary OOB computation when n_more_estimators == 0

#### Any other comments?
I understand from our previous conversation (https://github.com/scikit-learn/scikit-learn/pull/24579#discussion_r1012729686) that @glemaitre would like some unit tests for this change.

I noticed that the unit test below will check for the case where oob_score is toggled from `False` to `True`
https://github.com/scikit-learn/scikit-learn/blob/c5f10c8b51003caa1feb90b1c6cdd3f73615e45f/sklearn/ensemble/tests/test_forest.py#L1418-L1465

However, I am not sure how I should go about checking that OOB computation (call to `self._set_oob_score_and_attributes()` at https://github.com/scikit-learn/scikit-learn/blob/c5f10c8b51003caa1feb90b1c6cdd3f73615e45f/sklearn/ensemble/_forest.py#L492-L494 or https://github.com/scikit-learn/scikit-learn/blob/c5f10c8b51003caa1feb90b1c6cdd3f73615e45f/sklearn/ensemble/_forest.py#L496) is not called when `n_more_estimators == 0`.

I thought of checking on some object attribute that might be changed by `self._set_oob_score_and_attributes()` but I couldn't find any candidates. I did some research online and it seems like others create a mock function if they want to check for functions being called in runtime.

Do you have any suggestions on how I can go about writing the test for this?
<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

Commit:
Edited docs for warm_start and removed OOB computation for n_estimators == 0

Commit:
Removed change to OOB computation

Commit:
Removed change to OOB computation

Commit:
Removed OOB computation when n_more_estimators == 0

Commit:
Removed OOB computation when n_more_estimators == 0

thomasjpfan reviewed (commented):
Thank you for the PR @choo8 !

> Do you have any suggestions on how I can go about writing the test for this?

As for testing, I think we can use a [Python's Mock](https://docs.python.org/3/library/unittest.mock.html#quick-guide) object with a callable as the `side_effect`, pass it to `self.oob_score`, and assert the number of times it is called. With this PR, `self.oob_score` should be called less than on `main`.

Commit:
Added test to check for unnecessary OOB computation when n_more_estimators == 0

choo8 commented:
Thanks for the suggestion and comments @thomasjpfan! I've commited my test, please let me know if you have any comments. I decided to add on to the "check_warm_start_oob" test function as I thought the test logic should belong there.

thomasjpfan commented on sklearn/ensemble/tests/test_forest.py:
```diff
@@ -1456,13 +1457,24 @@ def check_warm_start_oob(name):
 
-    est_3.set_params(oob_score=True)
-    ignore_warnings(est_3.fit)(X, y)
+    # Patch _set_oob_score_and_attributes() to track OOB computation
```
I think extending this test complicates the original test a little too much. I prefer adding another test:

```python
def test_oob_not_computed_twice(monkeypatch):
    """Check that oob_score is not computed twice when warm_start=True."""
    est = RandomForestClassifier(n_estimators=10, oob_score=True, warm_start=True)

    mock = Mock(side_effect=est._set_oob_score_and_attributes)
    monkeypatch.setattr(est, "_set_oob_score_and_attributes", mock)

    est.fit(X, y)

    with pytest.warns(UserWarning, match="Warm-start fitting without increasing"):
        est.fit(X, y)

    mock.assert_called_once()
```

where `from unittest.mock import Mock`.

thomasjpfan commented on sklearn/ensemble/_forest.py:
```diff
@@ -357,11 +357,9 @@ def fit(self, X, y, sample_weight=None):
-                    " expected. Please change the shape of y to "
-                    "(n_samples,), for example using ravel()."
-                ),
+                "A column-vector y was passed when a 1d array was"
```
A few months ago we updated to `black==23.3.0`. May you update to `black==23.3.0` and run the linter again?

thomasjpfan reviewed (commented):
Thank you for the update!

Please add an entry to the change log at `doc/whats_new/v1.3.rst` with tag |Efficiency|. Like the other entries there, please reference this pull request with `:pr:` and credit yourself (and other contributors if applicable) with `:user:`.

Commit:
Split tests and updated changelog

choo8 commented:
Thank you for your comments @thomasjpfan. I've created a separate test case as suggested, updated the changelog and also updated the version of the linter.

Commit:
Corrected with linter

thomasjpfan commented:
There seems to be an issue with the merge because the diff is now at `+3,998 −1,450`. Can you try running the following:

```bash
git fetch upstream main
git rebase upstream/main
```

and do a force push?

choo8 commented:
Hi @thomasjpfan, I've fixed the diff with the commands you suggested.

thomasjpfan commented on doc/whats_new/v1.3.rst:
```diff
@@ -287,6 +287,10 @@ Changelog
   scikit-learn 1.3: retraining with scikit-learn 1.3 is required.
   :pr:`25186` by :user:`Felipe Breve Siola <fsiola>`.
 
+- |Efficiency| :class:`ensemble.BaseForest` now only recomputes out-of-bag scores
+  if `n_more_estimators > 0` in subsequent `fit` calls.
```
`BaseForest` is not a public class and `n_more_estimators` is not a public parameter. The user guide is public facing, so it's best to state the change in terms of the public API:

```suggestion
- |Efficiency| :class:`ensemble.RandomForestClassifier` and 
  :class:`ensemble.RandomForestRegressor` with `warm_start=True` now only
  recomputes out-of-bag scores when there are actually more `n_estimators`
  in subsequent `fit` calls.
```

thomasjpfan commented on sklearn/ensemble/tests/test_forest.py:
```diff
@@ -1470,6 +1471,31 @@ def test_warm_start_oob(name):
     check_warm_start_oob(name)
 
 
+def check_oob_not_computed_twice(name):
```
I can see we there is a convention in this file of having a separate function that the test calls, but I think that adds more indirection. Can you place the body of the function directly in the `test_oob_not_computed_twice` test?

thomasjpfan reviewed (commented):
Thank you for the update!

Commit:
Updated changelog and refactored unit test

choo8 commented:
Hi @thomasjpfan, I've updated the PR with regards to your latest comments.

thomasjpfan commented on sklearn/ensemble/tests/test_forest.py:
```diff
@@ -1466,8 +1467,29 @@ def check_warm_start_oob(name):
 
 
 @pytest.mark.parametrize("name", FOREST_CLASSIFIERS_REGRESSORS)
-def test_warm_start_oob(name):
```
Looks like something went wrong with the last patch.

Can you add `test_warm_start_oob` back and move the body of the new function into `test_oob_not_computed_twice`?

choo8 replied:
Sorry, I mistakenly deleted the wrong function. I've ran pytest to make sure and pushed the corrected edit.

Commit:
Fixed accidental deletion of old unit tests

thomasjpfan reviewed (approved):
Thank you for the update! LGTM

glemaitre commented on doc/whats_new/v1.3.rst:
```diff
@@ -287,6 +287,12 @@ Changelog
   scikit-learn 1.3: retraining with scikit-learn 1.3 is required.
   :pr:`25186` by :user:`Felipe Breve Siola <fsiola>`.
 
+- |Efficiency| :class:`ensemble.RandomForestClassifier` and 
```
```suggestion
- |Efficiency| :class:`ensemble.RandomForestClassifier` and
```

glemaitre reviewed (approved):
I will merge these two nitpicks.

glemaitre committed:
Apply suggestions from code review

glemaitre commented:
Thanks @choo8
